x-pack/filebeat/input/streaming/crowdstrike.go

```diff
@@ -0,0 +1,288 @@
+	var err error
+	for {
+		state, err = s.followSession(ctx, cli, state)
+		if err != nil {
+			if !errors.Is(err, Warning{}) {
+				if errors.Is(err, context.Canceled) {
+					return nil
+				}
+				s.metrics.errorsTotal.Inc()
+				return err
+			}
+			s.metrics.errorsTotal.Inc()
+			s.log.Warnw("session warning", "error", err)
+		}
+	}
+}
+
+func (s *falconHoseStream) followSession(ctx context.Context, cli *http.Client, state map[string]any) (map[string]any, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.discoverURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare discover stream request: %w", err)
+	}
+	resp, err := cli.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed GET to discover stream: %w", err)
+	}
+	defer resp.Body.Close()
+
+	dec := json.NewDecoder(resp.Body)
+
+	type resource struct {
+		FeedURL string `json:"dataFeedURL"`
+		Session struct {
+			Token   string    `json:"token"`
+			Expires time.Time `json:"expiration"`
+		} `json:"sessionToken"`
+		RefreshURL   string `json:"refreshActiveSessionURL"`
+		RefreshAfter int    `json:"refreshActiveSessionInterval"`
+	}
+	var body struct {
+		Resources []resource     `json:"resources"`
+		Meta      map[string]any `json:"meta"`
+	}
+	err = dec.Decode(&body)
+	if err != nil {
+		return state, Warning{fmt.Errorf("failed to decode discover body: %w", err)}
+	}
+	s.log.Debugw("stream discover metadata", "meta", mapstr.M(body.Meta))
+
+	var offset int
+	if cursor, ok := state["cursor"].(map[string]any); ok {
+		switch off := cursor["offset"].(type) {
+		case int:
+			offset = off
+		case float64:
+			offset = int(off)
+		}
+	}
+
+	for _, r := range body.Resources {
+		refreshAfter := time.Duration(r.RefreshAfter) * time.Second
+		go func() {
+			const grace = 5 * time.Minute
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(refreshAfter - grace):
+					req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.RefreshURL, nil)
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Errorw("failed to prepare refresh stream request", "error", err)
+						return
+					}
+					req.Header.Set("Content-Type", "application/json")
+					resp, err := cli.Do(req)
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Errorw("failed to refresh stream connection", "error", err)
+						return
+					}
+					err = resp.Body.Close()
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Warnw("failed to close refresh response body", "error", err)
+					}
+				}
+			}
+		}()
+
+		if offset > 0 {
+			feedURL, err := url.Parse(r.FeedURL)
+			if err != nil {
+				log.Fatalf("failed to parse feed url: %v", err)
+			}
+			feedQuery, err := url.ParseQuery(feedURL.RawQuery)
+			if err != nil {
+				log.Fatalf("failed to parse feed query: %v", err)
+			}
+			feedQuery.Set("offset", strconv.Itoa(offset))
+			feedURL.RawQuery = feedQuery.Encode()
+			r.FeedURL = feedURL.String()
+		}
```

Comment:
The calls to `log.Fatalf` on this block will cause Filebeat to exit instantly. An input failure should not crash the whole Filebeat instance.

If there is no way for the input to recover, returning an error, logging and stopping the input would be the best way to handle it.

Reply:
This is accidental leftovers from the original toy.